man/download-refseq-genome.1.ronn
# download-refseq-genome(1) - download refseq genome

## SYNOPSIS

`download-refseq-genome`
`--organism`=<organism>
[`--taxonomic-group`=<taxonomic_group>]
[`--genome-build`=<genome_build>]
[`--output-dir`=<output_dir>]

## DESCRIPTION

Download RefSeq genome and corresponding annotation files.

## OPTIONS

  * `organism`=<organism>:
    Full Latin organism name (e.g. `"Homo sapiens"`).

  * `taxonomic-group`=<taxonomic_group>:
    NCBI FTP server taxnomic group subdirectory path
    (e.g. `"vertebrate_mammalian"`).
    Defining this manually avoids having to query the FTP server.

  * `--genome-build`=<build>:
    Genome build (e.g. `"GCF_000001405.39_GRCh38.p12"`).
    Defaults to current genome build.

  * `--output-dir`=<output_dir>
    Output directory.
    Defaults to working directory.

## SEE ALSO

Primary RefSeq URLs:

- https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/
- ftp://ftp.ncbi.nlm.nih.gov/refseq/
- ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/

Current GRCh38 assembly:
ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/

- DNA: GRCh38_latest_genomic.fna.gz
- RNA: GRCh38_latest_rna.fna.gz
- GTF: GRCh38_latest_genomic.gtf.gz
- GFF: GRCh38_latest_genomic.gff.gz

Preformatted for analysis pipelines (UCSC IDs):
ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/

- Full:   GCA_000001405.15_GRCh38_full_analysis_set.fna.gz
- No ALT: GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz

Full version information in URL:
ftp://ftp.ncbi.nlm.nih.gov/genomes/Homo_sapiens/109.20190905/GCF_000001405.39_GRCh38.p13/

- DNA: GCF_000001405.39_GRCh38.p13_genomic.fna.gz
- RNA: GCF_000001405.39_GRCh38.p13_rna.fna.gz
- GTF: GCF_000001405.39_GRCh38.p13_genomic.gtf.gz
- GFF: GCF_000001405.39_GRCh38.p13_genomic.gff.gz

Alternate paths:

- ftp://ftp.ncbi.nlm.nih.gov/genomes/Homo_sapiens/
- ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/

## NOTE

R function, defined in r-acidgenomes package.


NEWS
## koopa 0.3.2 (2019-04-18)

- Improved build scripts, and hardened some intended for RedHat Linux.
- Updated genome build information.
- Improved bcbio scripts and custom genome builds.
- New exported scripts: `ip-address.sh`, `vim-sort.sh`.
- Added script to build Python from source for RedHat.



## koopa 0.3.1 (2019-03-19)

- Updated to Ensembl release to 95.
- Updated GENCODE to release human 29 / mouse M20.
- Updated FlyBase to 2019_01.
- `git-pull-all.sh`: Added `git fetch` call.
- Improved build configuration scripts: emacs, git, gsl, openssl, tmux.
- Added bcbio system configuration files (bcbio_system.yaml).
- Improved default run template for bcbio RNA-seq.



## koopa 0.3.0 (2019-02-20)

This release overhauls the previous activation method. Refer to the README
for details on how to reconfigure your shell profile.

- Changed activation method to use `activate.sh` script at top of repo.
- Improved support for ksh and zsh shells. Added Travis CI checks.
- Added native support for zsh pure prompt, including async script. These files
  are saved in `zsh/fpath/`.
- Reorganized bcbio configuration scripts and templates.
- Completely overhauled system activation method. Additional shell
  configuration, including prompt string customization is now disabled by
  default. Refer to the installation instructions for details on how to enable
  this new optional `extra` mode.
- Reorganized installer scripts exported in `bin/` including conda scripts,
  `reset-permissions.sh`, `sudo-yum-update.sh`, etc.



## koopa 0.2.8 (2019-02-12)

- Migrated darwin (macOS)-specific scripts from bash repo to koopa.
- Migrated Harvard O2 and Harvard Odyssey HPC scripts here from bash repo.
- Added `rsync-azure-files.sh` script for working with Azure Shares (Samba).
- Added `autopad-samples.sh` script.
- Added `bash-strict-mode.sh` script.
- Added `md5sum.sh` script.
- Reorganized other scripts accessible via `bin/`.
- Reorganized activation and system scripts, in preparation for 0.3 release.



## koopa 0.2.7 (2019-02-04)

- Improved shellchecks on workflows.
- Split out some activation steps into separate scripts.



## koopa 0.2.6 (2019-01-23)

- Added back ".sh" extension for all exported scripts.
- Updated Travis CI configuration to use shellcheck.
- Reorganized and improved workflow scripts.



## koopa 0.2.5 (2019-01-07)

- Updated installation instructions to recommend cloning to `~/.koopa`.
- Conda now will only activate when `CONDA_DEFAULT_ENV` is set in environment.
  This improves handling in situations for user accounts where we don't want
  to activate conda in a shared environemnt (e.g. bioinfo account on Azure).



## koopa 0.1.1 (2018-09-05)

- Now exporting `KOOPA_VERSION` and `KOOPA_DATE` global variables.
- Updated installation instructions.
- Removed `/setup/` config files.
- Added support for automatic SSH key import, using `$SSH_KEY` variable.
- Removed "g*" aliases for GitHub commands (e.g. ga, gc, gd). Too confusing.
- Simplified the separator bar for koopa message, avoiding unicode.
- Renamed `bcl2fastq_indrop` script to `bcl2fastq_indrops`.
- Added some utility scripts from macOS bash repo:
      - delete_fs_cruft
      - git_pull_all
      - gzip_dir
      - install_tex
      - move_files_up_1_level
      - rename_from_csv
      - reset_permissions



## koopa 0.1.0 (2018-08-24)

- Initial stable release.
- Repo was previously named "seqcloud".


lang/python/include/download-refseq-genome.py
#!/usr/bin/env python3
"""
Download RefSeq genome.
"""

from argparse import ArgumentParser
from os.path import join, realpath

from koopa.arg import dir_path
from koopa.files import download
from koopa.genome.refseq import (
    download_refseq_genome,
    download_refseq_gff,
    download_refseq_gtf,
    download_refseq_transcriptome,
    refseq_version,
)
from koopa.strings import paste_url
from koopa.syntactic import kebab_case
from koopa.system import koopa_help

koopa_help()


parser = ArgumentParser()
parser.add_argument(
    "--build",
    default="GRCh38",
    const="GRCh38",
    nargs="?",
    choices=["GRCh38", "GRCh37"],
)
parser.add_argument(
    "--type",
    default="all",
    const="all",
    nargs="?",
    choices=["all", "genome", "transcriptome", "none"],
)
parser.add_argument(
    "--annotation",
    default="all",
    const="all",
    nargs="?",
    choices=["all", "gtf", "gff", "none"],
)
parser.add_argument("--output-dir", type=dir_path, default=".")
parser.add_argument("--decompress", action="store_true")
args = parser.parse_args()


def main(annotation, build, decompress, genome_type, output_dir):
    """
    Download RefSeq genome.
    """
    organism = "Homo sapiens"
    organism_short = "H_sapiens"
    release = refseq_version()
    release_url = paste_url(
        "ftp://ftp.ncbi.nlm.nih.gov",
        "refseq",
        organism_short,
        "annotation",
        build + "_latest",
        "refseq_identifiers",
    )
    output_basename = kebab_case(
        organism + " " + build + " " + "refseq" + " " + release
    )
    output_dir = join(realpath(output_dir), output_basename)
    download(url=paste_url(release_url, "README.txt"), output_dir=output_dir)
    download(
        url=paste_url(release_url, build + "_latest_assembly_report.txt"),
        output_dir=output_dir,
    )
    if genome_type == "genome":
        download_refseq_genome(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif genome_type == "transcriptome":
        download_refseq_transcriptome(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif genome_type == "all":
        download_refseq_genome(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
        download_refseq_transcriptome(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    if annotation == "gtf":
        download_refseq_gtf(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif annotation == "gff":
        download_refseq_gff(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif annotation == "all":
        download_refseq_gtf(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
        download_refseq_gff(
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    print("Genome downloaded successfully to '" + output_dir + "'.")


main(
    annotation=args.annotation,
    build=args.build,
    decompress=args.decompress,
    genome_type=args.type,
    output_dir=args.output_dir,
)
